File: Recursion/Leetcode/Medium/CombinationSum_II/CombinationSum_II.java
// https://leetcode.com/problems/combination-sum-ii/
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSum_II {
    void generate(int[] candidates, int target, int start, List<Integer> seq, List<List<Integer>> combo) {
        if (start == candidates.length || target <= 0) {
            if (target == 0) {
                combo.add(new ArrayList<>(seq));
            }
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            if (i != start && candidates[i] == candidates[i - 1]) {
                continue;
            }
            if (target < candidates[i]) {
                break;
            }
            seq.add(candidates[i]);
            generate(candidates, target - candidates[i], i + 1, seq, combo);
            seq.remove(seq.size() - 1);
        }
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> list = new ArrayList<>();
        List<Integer> seq = new ArrayList<>();
        generate(candidates, target, 0, seq, list);
        return list;
    }
    public static void main(String[] args) {
        
    }
}

File: Recursion/Leetcode/Medium/CombinationSum_II/CombinationSum_II.cpp
// https://leetcode.com/problems/combination-sum-ii/
#include<bits/stdc++.h>
using namespace std;
void combinations2(vector<int> &candidates, int target, vector<int> &v, vector<vector<int>> &combo, int start)
{
    if (target == 0)
    {
        combo.emplace_back(v);
        return;
    }
    for (int i = start; i < candidates.size(); i++)
    {
        if (i != start && candidates[i] == candidates[i - 1])
            continue;
        if (candidates[i] > target)
            break;
        v.emplace_back(candidates[i]);
        combinations2(candidates, target - candidates[i], v, combo, i + 1);
        v.pop_back();
    }
}
vector<vector<int>> combinationSum2(vector<int> &candidates, int target)
{
    sort(candidates.begin(), candidates.end());
    vector<vector<int>> arr;
    vector<int> v;
    combinations2(candidates, target, v, arr, 0);
    return arr;
}
int main(void) {
    
    return 0;
}